### PR TITLE
Resolve method reference receivers

### PR DESCRIPTION
**Pull request by vinistock**
This PR starts resolving method reference receivers, so that the expected singleton classes are created and get their ancestors linearized.

The idea is this:

```ruby
Foo.bar # this is `Foo::<Foo>#bar` and therefore we need <Foo> to exist
Foo.singleton_class.bar # this is `Foo::<Foo>::<<Foo>>#bar` and so we need both <Foo> and <<Foo>> to exist
```

For this reason, we need to loop to create all singleton levels involved in the method reference.

### Trade offs

The resolution time added is actually not terrible, but memory usage is a different story.

This PR stack will eliminate the need to mutate the graph after resolution, but memory usage increases in Core by 1.5 GB.

### CODE REVIEW

File: rust/rubydex/src/resolution.rs [hunk 1]
```diff
@@ -132,12 +134,15 @@ impl<'a> Resolver<'a> {
...
-                    Unit::Reference(id) => {
+                    Unit::ConstantRef(id) => {
                         self.handle_reference_unit(&mut unit_queue, &mut made_progress, unit_id, id);
                     }
                     Unit::Ancestors(id) => {
                         self.handle_ancestor_unit(&mut unit_queue, &mut made_progress, id);
                     }
+                    Unit::MethodRef(id) => {
+                        self.handle_method_reference_unit(&mut unit_queue, &mut made_progress, id);
+                    }
```
Comment (Morriar) [inline rust/rubydex/src/resolution.rs]: See my comment https://github.com/Shopify/rubydex/pull/530/changes#r2742491771, I was wondering if the receiver could be automatically resolved through a constant reference with `Unit::ConstantRef`

File: rust/rubydex/src/resolution.rs [hunk 2]
```diff
@@ -4189,4 +4279,163 @@ mod tests {
...
+        });
+        context.resolve();
+
+        let singleton_foo = context
+            .graph()
+            .declarations()
+            .get(&DeclarationId::from("Foo::<Foo>::<<Foo>>::<<<Foo>>>"))
+            .unwrap();
+
+        assert_eq!(1, singleton_foo.references().len());
```
Comment (Morriar) [inline rust/rubydex/src/resolution.rs]:
```suggestion
        assert_references_to!("Foo::<Foo>::<<Foo>>::<<<Foo>>>", 1)
```

File: rust/rubydex/src/resolution.rs (hunk 1 above)
Comment (vinistock) [reply rust/rubydex/src/resolution.rs]: You were actually right! I found a way to not have this new method and incorporate everything into constant resolution.

File: rust/rubydex/src/resolution.rs [hunk 3]
```diff
@@ -4089,4 +4097,135 @@ mod tests {
...
+
+    #[test]
+    fn method_calls_automatically_trigger_singleton_creation() {
+        let mut context = GraphTest::new();
+        context.index_uri("file:///foo.rb", {
+            "
+            class Foo
+            end
+
+            Foo.
```
Comment (st0012) [inline rust/rubydex/src/resolution.rs:4109]: Do we need to simulate method completion situation in rubydex tests? This is incorrect syntax and IMO is confusing to the reader. `Foo.bar` should probably work too?

Comment (st0012) [reply rust/rubydex/src/resolution.rs:4109]: Also, should we have `assert_no_diagnostics` in the new tests?

File: rust/rubydex/src/model/name.rs [hunk 4]
```diff
@@ -13,6 +13,9 @@ pub enum ParentScope {
     TopLevel,
     /// There's a parent scope in this reference (e.g.: `Foo::Bar`)
     Some(NameId),
+    /// Parent scope representing an attached singleton class (e.g., `<Foo>` in `Foo::<Foo>`).  Unlike `Some`, which
+    /// triggers ancestor search during resolution, `Attached` triggers singleton class creation.
+    Attached(NameId),
```
Comment (Morriar) [inline rust/rubydex/src/model/name.rs:21]: This name is confusing because it's the opposite of Sorbet `attached_class`: https://sorbet.org/docs/attached-class

Why not `Singleton`?

File: rust/rubydex/src/resolution.rs (hunk 3 above)
Comment (vinistock) [reply rust/rubydex/src/resolution.rs:4109]: The thing is, if we don't want to allow lazy creation of singletons, we have to ensure that `Foo.` is properly creating the singleton entity or else completion won't work.

I think we have to keep this test or else it's easy to forget this particular use case.

File: rust/rubydex/src/test_utils/graph_test.rs [hunk 5]
```diff
@@ -465,6 +465,21 @@ macro_rules! assert_no_diagnostics {
     }};
 }
 
+/// Asserts the number of references registered to a specific declaration
+#[cfg(test)]
+#[macro_export]
+macro_rules! assert_reference_count {
```
Comment (Morriar) [inline rust/rubydex/src/test_utils/graph_test.rs]: Looks like you're duplicating https://github.com/Shopify/rubydex/blob/main/rust/rubydex/src/test_utils/graph_test.rs#L249. I'm open to a renaming if you want.

File: rust/rubydex/src/model/name.rs (hunk 4 above)
Comment (vinistock) [reply rust/rubydex/src/model/name.rs:21]: Maybe the comment is confusing because that's not what it means. This is what it means (which matches Sorbet):

```
Foo::<Foo>
^ ParentScope::Attached
      ^ Just a name

Foo::<Foo>::<<Foo>>
^ Attached for <Foo>
      ^ Attached for <<Foo>>
```

File: rust/rubydex/src/resolution.rs [hunk 6]
```diff
@@ -4089,4 +4097,135 @@ mod tests {
...
+        context.resolve();
+
+        assert!(
+            context
+                .graph()
+                .declarations()
+                .get(&DeclarationId::from("Foo::<Foo>"))
+                .is_none()
+        );
+    }
```
Comment (Morriar) [inline rust/rubydex/src/resolution.rs]:
```suggestion
        assert_declaration_does_not_exist!("Foo::<Foo>");
    }
```

File: rust/rubydex/src/model/name.rs (hunk 4 above)
Comment (Morriar) [reply rust/rubydex/src/model/name.rs:21]: Yeah the comment got me confused.

Maybe something like:

```rs
/// Parent scope representing the class attached to a singleton class context
///
///  Foo::<Foo>::<<Foo>>
///  ^ Attached for <Foo>
///        ^ Attached for <<Foo>>
```

File: rust/rubydex/src/resolution.rs (hunk 3 above)
Comment (st0012) [reply rust/rubydex/src/resolution.rs:4109]: So the behaviour between seeing `Foo.` and `Foo.b` is different? Can you point me to the code that makes the distinction?

Comment (vinistock) [reply rust/rubydex/src/resolution.rs:4109]: It's not different, the point is we want to ensure that it's not. In `Foo.b`, you have no syntax errors and a proper method call. This scenario is naturally expected to create the singleton class because you're invoking a singleton method.

`Foo.` on the other hand is a bit of an edge case. You have an incomplete method call for which the message is empty. The tests is just to ensure that even in this case we still produce the singleton class, otherwise we can't provide completion.

File: rust/rubydex/src/test_utils/graph_test.rs (hunk 5 above)
Comment (vinistock) [reply rust/rubydex/src/test_utils/graph_test.rs]: Nah, it's all good. I just hadn't found it.

File: rust/rubydex/src/resolution.rs (hunk 3 above)
Comment (vinistock) [reply rust/rubydex/src/resolution.rs:4109]: Added the diagnostic assertions.

Comment (st0012) [reply rust/rubydex/src/resolution.rs:4109]: Can we indicate this intention in the test name and a comment on this line? Something like `incomplete_method_calls_automatically_trigger_singleton_creation`

Comment (vinistock) [reply rust/rubydex/src/resolution.rs:4109]: Renamed it 👍
